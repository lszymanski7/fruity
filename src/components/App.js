/* eslint-disable react-hooks/exhaustive-deps */

import React, { useState, useEffect } from 'react'
import Button from './Button'
import Footer from './Footer'
import Header from './Header'
import Modal from './Modal'
import Widget from './Widget'
import { fruits } from '../data/constants'

const App = () => {
    const storage = localStorage.getItem('fruits')
    const initialValue = storage === null ? fruits : JSON.parse(storage)
    const [data, setData] = useState(initialValue)
    const [selected, setSelected] = useState(null)
    const [isOpen, setIsOpen] = useState(false)
    const filteredData = data.filter((fruit) => fruit.checked === true)

    const openModal = (fruit) => {
        setSelected(fruit)
        setIsOpen(true)
    }

    const closeModal = () => {
        setSelected(null)
        setIsOpen(false)
    }

    const handleCheck = (id) => {
        const newData = data.map((fruit) => ({ ...fruit }))
        const index = newData.findIndex((fruit) => fruit.id === id)
        newData[index].checked = !newData[index].checked
        setData(newData)
    }

    const handleDraw = () => {
        const rand = Math.floor(Math.random() * filteredData.length)
        openModal(filteredData[rand])
    }

    const handleReset = () => {
        const newData = data.map((fruit) => ({ ...fruit, checked: false }))
        setData(newData)
    }

    useEffect(() => {
        if (storage === null) {
            localStorage.setItem('fruits', JSON.stringify(fruits))
        }
    }, [])

    useEffect(() => {
        localStorage.setItem('fruits', JSON.stringify(data))
    }, [data])

    return (
        <div className="flexbox-fullscreen">
            <Header />
            <div className="flexbox-container">
                <Widget
                    data={data}
                    handleCheck={handleCheck}
                    handleReset={handleReset}
                />
                <Button
                    disabled={filteredData.length < 2}
                    handleDraw={handleDraw}
                />
            </div>
            <Footer />
            <Modal
                isOpen={isOpen}
                onRequestClose={closeModal}
                selected={selected}
            />
        </div>
    )
}

export { App as default }
