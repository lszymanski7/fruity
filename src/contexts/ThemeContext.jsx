/* istanbul ignore file */

import React, { createContext, useEffect } from 'react'
import PropTypes from 'prop-types'
import useLocalStorage from '../hooks/useLocalStorage'

// Function ➞ creates theme context
export const ThemeContext = createContext()

// Function ➞ creates theme provider
export const ThemeProvider = ({ children }) => {
  const [theme, setTheme] = useLocalStorage('theme', 'light')

  const toggleTheme = () => {
    const nextState = theme === 'dark' ? 'light' : 'dark'
    setTheme(nextState)
  }

  useEffect(() => {
    switch (theme) {
      case 'dark':
        document.body.className = 'theme-dark'
        break
      case 'light':
        document.body.className = 'theme-light'
        break
      default:
        throw new Error(`Unhandled theme: ${theme}`)
    }
  }, [theme])

  return (
    <ThemeContext.Provider value={{ theme, toggleTheme }}>
      {children}
    </ThemeContext.Provider>
  )
}

ThemeProvider.propTypes = {
  children: PropTypes.object.isRequired
}
