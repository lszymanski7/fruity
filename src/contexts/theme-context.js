/* istanbul ignore file */

import React, { createContext, useEffect } from 'react'
import PropTypes from 'prop-types'
import useLocalStorage from '../hooks/useLocalStorage'

// Function ➞ creates theme context
const ThemeContext = createContext()

// Function ➞ creates theme provider
const ThemeProvider = ({ children }) => {
    const [theme, setTheme] = useLocalStorage('theme', 'light')

    const toggleTheme = () => {
        const nextState = theme === 'dark' ? 'light' : 'dark'
        setTheme(nextState)
    }

    useEffect(() => {
        switch (theme) {
            case 'dark':
                document.body.className = 'background-dark'
                break
            case 'light':
                document.body.className = 'background-light'
                break
            default:
                throw new Error(`Unhandled theme: ${theme}`)
        }
    }, [theme])

    return (
        <ThemeContext.Provider value={{ theme, toggleTheme }}>
            {children}
        </ThemeContext.Provider>
    )
}

ThemeProvider.propTypes = {
    children: PropTypes.object.isRequired
}

export { ThemeContext, ThemeProvider }
