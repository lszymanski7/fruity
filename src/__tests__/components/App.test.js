import React from 'react'
import { render } from '@testing-library/react'
import App from '../../components/App'
import { ThemeProvider } from '../../contexts/theme-context'

describe('App', () => {
    it('should be rendered correctly.', () => {
        const { container } = render(<App />, { wrapper: ThemeProvider })
        expect(container).toMatchSnapshot()
    })
})
