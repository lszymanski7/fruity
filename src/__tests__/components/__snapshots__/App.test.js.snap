// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should be rendered correctly. 1`] = `
<div>
  <div
    class="fullscreen"
  >
    <header
      class="header"
    >
      <div
        class="header__widgets-container"
      >
        <a
          href="https://github.com/lszymanski7/fruity-app"
        >
          <img
            alt="GitHub Logo"
            class="header__github-logo header__github-logo--light"
            src="github.svg"
          />
        </a>
        <div
          class="toggle-switch"
        >
          <input
            class="toggle-switch__input"
            id="theme-switcher"
            type="checkbox"
          />
          <label
            class="toggle-switch__label"
            for="theme-switcher"
          >
            <span
              class="toggle-switch__icon moon-icon"
            />
            <span
              class="toggle-switch__icon sun-icon"
            />
            <span
              class="toggle-switch__slider"
            />
          </label>
        </div>
      </div>
      <div
        class="header__brand-container"
      >
        <img
          alt="Application Logo"
          class="header__app-logo"
          src="fruity-256x256.webp"
        />
        <h1
          class="header__app-name"
        >
          FRUITY
        </h1>
      </div>
    </header>
    <div
      class="board"
    >
      <div
        class="board__elements-container"
      >
        <div
          class="element"
        >
          <label
            class="element__label"
          >
            <input
              class="element__checkbox"
              type="checkbox"
            />
            <img
              alt="apple"
              class="element__image"
              src="apple.webp"
            />
            <span
              class="element__caption"
            >
              APPLE
            </span>
          </label>
        </div>
        <div
          class="element"
        >
          <label
            class="element__label"
          >
            <input
              class="element__checkbox"
              type="checkbox"
            />
            <img
              alt="banana"
              class="element__image"
              src="banana.webp"
            />
            <span
              class="element__caption"
            >
              BANANA
            </span>
          </label>
        </div>
        <div
          class="element"
        >
          <label
            class="element__label"
          >
            <input
              class="element__checkbox"
              type="checkbox"
            />
            <img
              alt="blackberry"
              class="element__image"
              src="blackberry.webp"
            />
            <span
              class="element__caption"
            >
              BLACKBERRY
            </span>
          </label>
        </div>
        <div
          class="element"
        >
          <label
            class="element__label"
          >
            <input
              class="element__checkbox"
              type="checkbox"
            />
            <img
              alt="blueberries"
              class="element__image"
              src="blueberries.webp"
            />
            <span
              class="element__caption"
            >
              BLUEBERRIES
            </span>
          </label>
        </div>
        <div
          class="element"
        >
          <label
            class="element__label"
          >
            <input
              class="element__checkbox"
              type="checkbox"
            />
            <img
              alt="cherries"
              class="element__image"
              src="cherries.webp"
            />
            <span
              class="element__caption"
            >
              CHERRIES
            </span>
          </label>
        </div>
        <div
          class="element"
        >
          <label
            class="element__label"
          >
            <input
              class="element__checkbox"
              type="checkbox"
            />
            <img
              alt="coconut"
              class="element__image"
              src="coconut.webp"
            />
            <span
              class="element__caption"
            >
              COCONUT
            </span>
          </label>
        </div>
        <div
          class="element"
        >
          <label
            class="element__label"
          >
            <input
              class="element__checkbox"
              type="checkbox"
            />
            <img
              alt="dragon fruit"
              class="element__image"
              src="dragon-fruit.webp"
            />
            <span
              class="element__caption"
            >
              DRAGON FRUIT
            </span>
          </label>
        </div>
        <div
          class="element"
        >
          <label
            class="element__label"
          >
            <input
              class="element__checkbox"
              type="checkbox"
            />
            <img
              alt="fig"
              class="element__image"
              src="fig.webp"
            />
            <span
              class="element__caption"
            >
              FIG
            </span>
          </label>
        </div>
        <div
          class="element"
        >
          <label
            class="element__label"
          >
            <input
              class="element__checkbox"
              type="checkbox"
            />
            <img
              alt="grapes"
              class="element__image"
              src="grapes.webp"
            />
            <span
              class="element__caption"
            >
              GRAPES
            </span>
          </label>
        </div>
        <div
          class="element"
        >
          <label
            class="element__label"
          >
            <input
              class="element__checkbox"
              type="checkbox"
            />
            <img
              alt="kiwi"
              class="element__image"
              src="kiwi.webp"
            />
            <span
              class="element__caption"
            >
              KIWI
            </span>
          </label>
        </div>
        <div
          class="element"
        >
          <label
            class="element__label"
          >
            <input
              class="element__checkbox"
              type="checkbox"
            />
            <img
              alt="lemon"
              class="element__image"
              src="lemon.webp"
            />
            <span
              class="element__caption"
            >
              LEMON
            </span>
          </label>
        </div>
        <div
          class="element"
        >
          <label
            class="element__label"
          >
            <input
              class="element__checkbox"
              type="checkbox"
            />
            <img
              alt="lychee"
              class="element__image"
              src="lychee.webp"
            />
            <span
              class="element__caption"
            >
              LYCHEE
            </span>
          </label>
        </div>
        <div
          class="element"
        >
          <label
            class="element__label"
          >
            <input
              class="element__checkbox"
              type="checkbox"
            />
            <img
              alt="mango"
              class="element__image"
              src="mango.webp"
            />
            <span
              class="element__caption"
            >
              MANGO
            </span>
          </label>
        </div>
        <div
          class="element"
        >
          <label
            class="element__label"
          >
            <input
              class="element__checkbox"
              type="checkbox"
            />
            <img
              alt="melon"
              class="element__image"
              src="melon.webp"
            />
            <span
              class="element__caption"
            >
              MELON
            </span>
          </label>
        </div>
        <div
          class="element"
        >
          <label
            class="element__label"
          >
            <input
              class="element__checkbox"
              type="checkbox"
            />
            <img
              alt="orange"
              class="element__image"
              src="orange.webp"
            />
            <span
              class="element__caption"
            >
              ORANGE
            </span>
          </label>
        </div>
        <div
          class="element"
        >
          <label
            class="element__label"
          >
            <input
              class="element__checkbox"
              type="checkbox"
            />
            <img
              alt="papaya"
              class="element__image"
              src="papaya.webp"
            />
            <span
              class="element__caption"
            >
              PAPAYA
            </span>
          </label>
        </div>
        <div
          class="element"
        >
          <label
            class="element__label"
          >
            <input
              class="element__checkbox"
              type="checkbox"
            />
            <img
              alt="peach"
              class="element__image"
              src="peach.webp"
            />
            <span
              class="element__caption"
            >
              PEACH
            </span>
          </label>
        </div>
        <div
          class="element"
        >
          <label
            class="element__label"
          >
            <input
              class="element__checkbox"
              type="checkbox"
            />
            <img
              alt="pear"
              class="element__image"
              src="pear.webp"
            />
            <span
              class="element__caption"
            >
              PEAR
            </span>
          </label>
        </div>
        <div
          class="element"
        >
          <label
            class="element__label"
          >
            <input
              class="element__checkbox"
              type="checkbox"
            />
            <img
              alt="pineapple"
              class="element__image"
              src="pineapple.webp"
            />
            <span
              class="element__caption"
            >
              PINEAPPLE
            </span>
          </label>
        </div>
        <div
          class="element"
        >
          <label
            class="element__label"
          >
            <input
              class="element__checkbox"
              type="checkbox"
            />
            <img
              alt="plum"
              class="element__image"
              src="plum.webp"
            />
            <span
              class="element__caption"
            >
              PLUM
            </span>
          </label>
        </div>
        <div
          class="element"
        >
          <label
            class="element__label"
          >
            <input
              class="element__checkbox"
              type="checkbox"
            />
            <img
              alt="pomegranate"
              class="element__image"
              src="pomegranate.webp"
            />
            <span
              class="element__caption"
            >
              POMEGRANATE
            </span>
          </label>
        </div>
        <div
          class="element"
        >
          <label
            class="element__label"
          >
            <input
              class="element__checkbox"
              type="checkbox"
            />
            <img
              alt="star fruit"
              class="element__image"
              src="star-fruit.webp"
            />
            <span
              class="element__caption"
            >
              STAR FRUIT
            </span>
          </label>
        </div>
        <div
          class="element"
        >
          <label
            class="element__label"
          >
            <input
              class="element__checkbox"
              type="checkbox"
            />
            <img
              alt="strawberry"
              class="element__image"
              src="strawberry.webp"
            />
            <span
              class="element__caption"
            >
              STRAWBERRY
            </span>
          </label>
        </div>
        <div
          class="element"
        >
          <label
            class="element__label"
          >
            <input
              class="element__checkbox"
              type="checkbox"
            />
            <img
              alt="watermelon"
              class="element__image"
              src="watermelon.webp"
            />
            <span
              class="element__caption"
            >
              WATERMELON
            </span>
          </label>
        </div>
      </div>
    </div>
    <div
      class="draw"
    >
      <div
        class="draw__buttons-container"
      >
        <button
          class="draw__start-button"
          disabled=""
          type="button"
        >
          SELECT MORE FRUITS
        </button>
        <button
          aria-label="Reset"
          class="draw__reset-button"
          type="button"
        >
          <img
            alt="Reset Icon"
            class="draw__reset-icon"
            src="reset.svg"
          />
        </button>
      </div>
    </div>
  </div>
</div>
`;
